transport/src/main/java/io/netty/channel/DelegatingChannelPromiseNotifier.java

```diff
@@ -0,0 +1,226 @@
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+@UnstableApi
+public final class DelegatingChannelPromiseNotifier implements ChannelPromise, ChannelFutureListener {
+    private static final InternalLogger logger =
+            InternalLoggerFactory.getInstance(DelegatingChannelPromiseNotifier.class);
+    private final ChannelPromise delegate;
+    private final boolean logNotifyFailure;
+
+    public DelegatingChannelPromiseNotifier(ChannelPromise delegate) {
+        this(delegate, true);
+    }
+
+    public DelegatingChannelPromiseNotifier(ChannelPromise delegate, boolean logNotifyFailure) {
+        this.delegate = checkNotNull(delegate, "delegate");
+        this.logNotifyFailure = logNotifyFailure;
+    }
+
+    @Override
+    public void operationComplete(ChannelFuture future) throws Exception {
+        InternalLogger internalLogger = logNotifyFailure ? logger : null;
+        if (future.isSuccess()) {
+            Void result = future.get();
+            PromiseNotificationUtil.trySuccess(delegate, result, internalLogger);
+        } else if (future.isCancelled()) {
+            PromiseNotificationUtil.tryCancel(delegate, internalLogger);
+        } else {
+            Throwable cause = future.cause();
+            PromiseNotificationUtil.tryFailure(delegate, cause, internalLogger);
+        }
+    }
+
+    @Override
+    public Channel channel() {
+        return delegate.channel();
+    }
+
+    @Override
+    public ChannelPromise setSuccess(Void result) {
+        delegate.setSuccess(result);
+        return this;
+    }
+
+    @Override
+    public ChannelPromise setSuccess() {
+        delegate.setSuccess();
+        return this;
+    }
+
+    @Override
+    public boolean trySuccess() {
+        return delegate.trySuccess();
+    }
+
+    @Override
+    public boolean trySuccess(Void result) {
+        return delegate.trySuccess(result);
+    }
+
+    @Override
+    public ChannelPromise setFailure(Throwable cause) {
+        delegate.setFailure(cause);
+        return this;
+    }
+
+    @Override
+    public ChannelPromise addListener(GenericFutureListener<? extends Future<? super Void>> listener) {
+        delegate.addListener(listener);
+        return this;
+    }
+
+    @Override
+    public ChannelPromise addListeners(GenericFutureListener<? extends Future<? super Void>>[] listeners) {
+        delegate.addListeners(listeners);
+        return this;
+    }
+
+    @Override
+    public ChannelPromise removeListener(GenericFutureListener<? extends Future<? super Void>> listener) {
+        delegate.removeListener(listener);
+        return this;
+    }
+
+    @Override
+    public ChannelPromise removeListeners(GenericFutureListener<? extends Future<? super Void>>[] listeners) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Consider using varargs for methods or constructors which take an array the last parameter. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AUseVarargs)